Add PlanetScale documentation (#1464)
brefphp/bref: docs/environment/database-planetscale.md

DATABASE_URL="mysql://<USERNAME>:<PASSWORD>@<HOST_URL>:3306/<DATABASE_NAME>?serverVersion=8.0"
```

If you don't deploy the `.env` file, you can configure the variables in `serverless.yml`:

```yaml
provider:
    # ...
    environment:
        DATABASE_URL: ${ssm:/my-app/database-url}
```

Note that the `DATABASE_URL` value is sensitive and can be set up as a secret via SSM. Read about [Secret variables](./variables.md#secrets) to learn more.

Finally, edit the `config/packages/doctrine.yaml` configuration file to set up [the SSL connections](https://planetscale.com/docs/concepts/secure-connections):

```yaml
doctrine:
    dbal:
        url: '%env(resolve:DATABASE_URL)%'
        options:
            # Connect to the database via SSL
            !php/const:PDO::MYSQL_ATTR_SSL_CA: /opt/bref/ssl/cert.pem

# ...
```

Let's deploy the changes:

```bash
serverless deploy
```

Now that Symfony is configured, we can run the `bin/console doctrine:migrations:migrate` command in AWS Lambda to set up our tables:

```bash
serverless bref:cli --args="doctrine:migrations:migrate"
```

That's it! Our database is ready to use.

## Digging deeper

### MySQL compatibility

### Database import

> ### Schema changes workflow

Author: Added the titles of extra sections I would like to cover.

The idea here would be to provide pointers and an overview (short), not duplicate the official docs. But these would be very common questions, and I'd love to point people in the right direction.

- [**Doctrine**](https://www.doctrine-project.org/) works fine, but you should not use [`onDelete="CASCADE"` which relies on foreign key constraints](https://www.doctrine-project.org/projects/doctrine-orm/en/2.14/reference/working-with-objects.html#removing-entities) (`cascade=REMOVE` or `cascade=ALL` is fine as the cascade is performed by Doctrine in memory).

You can read the [complete **MySQL compatibility table** on the PlanetScale website](https://planetscale.com/docs/reference/mysql-compatibility).

## Database import

PlanetScale provides an automated import tool to import an existing database without downtime. Check out [the documentation](https://planetscale.com/docs/imports/database-imports) to get started.

For simple scenarios, you can also use the [`mysqldump` tool](https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html#mysqldump-syntax) to export you existing database and import it later in PlanetScale. Note that there are [specific options we need to use for Vitess](https://vitess.io/docs/15.0/user-guides/configuration-basic/exporting-data/#mysqldump). First, let's export the existing database:

```bash

> mysqldump -u <user> -p<password> -h <hostname> --set-gtid-purged=OFF --no-tablespaces <db-name> > dump.sql

Author: TODO: sort out `mysqldump` options to deal with foreign keys.

introduction: Configure Bref to use a PlanetScale database in your PHP application on AWS Lambda.
---

[PlanetScale](https://planetscale.com/) is a hosted serverless MySQL database based on the Vitess engine ([learn more](https://planetscale.com/docs/concepts/what-is-planetscale)).

Amongst other features, it offers the following benefits compared to running a database on AWS:

- Simple to set up: no VPC (virtual private network) to set up, no instances to configure.
- Runs [the Vitess clustering system](https://planetscale.com/blog/vitess-for-the-rest-of-us), which offers better scalability and supports a lot more concurrent connections [via built-in connection pooling](https://planetscale.com/blog/one-million-connections).
- Offers a [free database in the Hobby plan](https://planetscale.com/pricing), and paid plans are usage-based.
- Since it does not require a VPC, we do not need to set up and pay [for a NAT Gateway](database.md#accessing-the-internet).

One extra feature worth mentioning is [the branching concept](https://planetscale.com/docs/concepts/branching): it enables testing schema changes before deploying them in production without downtime.

## Getting started

> To use PlaneScale with Bref, start [by creating a PlanetScale account](https://planetscale.com/).

Reviewer: ```suggestion
To use PlanetScale with Bref, start [by creating a PlanetScale account](https://planetscale.com/).
```

If you deploy a `.env` file, set up the following variables:

```bash
DB_CONNECTION=mysql
DB_HOST=<host url>
DB_PORT=3306
DB_DATABASE=<database name>
DB_USERNAME=<user>
DB_PASSWORD=<password>
# Connect via SSL (https://planetscale.com/docs/concepts/secure-connections)
MYSQL_ATTR_SSL_CA=/opt/bref/ssl/cert.pem
```

If you don't deploy the `.env` file, you can configure the variables in `serverless.yml`:

```yaml
provider:
    # ...
    environment:
        DB_HOST: <host url>
        DB_DATABASE: <database name>
        DB_USERNAME: <user>
        DB_PASSWORD: ${ssm:/my-app/database-password}
        # Connect via SSL (https://planetscale.com/docs/concepts/secure-connections)
        MYSQL_ATTR_SSL_CA: /opt/bref/ssl/cert.pem
```

Note that the `DB_PASSWORD` value is sensitive and can be set up as a secret via SSM. Read about [Secret variables](./variables.md#secrets) to learn more.

Don't forget to deploy the changes:

```bash
serverless deploy
```

Now that Laravel is configured, you can run `php artisan migrate` in AWS Lambda to set up our tables:

```bash
serverless bref:cli --args="migrate"
```

> 

Reviewer: ```suggestion
Note: PlanetScale does not support foreign key constraints. If your application uses foreign key constraints, you will have to remove them before running `php artisan migrate`. See the [MySQL compatibility section](#mysql-compatibility) for more information.

```

Reviewer: Might be worth it to add this note here too

If you deploy a `.env` file, set up the following variables:

```bash
DATABASE_URL="mysql://<USERNAME>:<PASSWORD>@<HOST_URL>:3306/<DATABASE_NAME>?serverVersion=8.0"
```

If you don't deploy the `.env` file, you can configure the variables in `serverless.yml`:

```yaml
provider:
    # ...
    environment:
        DATABASE_URL: ${ssm:/my-app/database-url}
```

Note that the `DATABASE_URL` value is sensitive and can be set up as a secret via SSM. Read about [Secret variables](./variables.md#secrets) to learn more.

Finally, edit the `config/packages/doctrine.yaml` configuration file to set up [the SSL connections](https://planetscale.com/docs/concepts/secure-connections):

```yaml
doctrine:
    dbal:
        url: '%env(resolve:DATABASE_URL)%'
        options:
            # Connect to the database via SSL
            !php/const:PDO::MYSQL_ATTR_SSL_CA: /opt/bref/ssl/cert.pem

# ...
```

Let's deploy the changes:

```bash
serverless deploy
```

Now that Symfony is configured, you can run the `bin/console doctrine:migrations:migrate` command in AWS Lambda to set up our tables:

```bash
serverless bref:cli --args="doctrine:migrations:migrate"
```

> 

Reviewer: ```suggestion
Note: PlanetScale does not support foreign key constraints. If your application uses foreign key constraints, you will have to remove them before running your migrations. See the [MySQL compatibility section](#mysql-compatibility) for more information.

```

That's it! Our database is ready to use.

## MySQL compatibility

> PlanetScale is based on the Vitess clustering system, which was built for scaling MySQL. Because of that, Vitess only supports some of the usual MySQL features.

Reviewer: ```suggestion
PlanetScale is based on the Vitess clustering system, which was built for scaling MySQL. Because of that, Vitess doesn't support all MySQL features.
```

```bash
mysqldump -u <user> -p<password> -h <hostname> --set-gtid-purged=OFF --no-tablespaces --no-data <db-name> > schema.sql
mysqldump -u <user> -p<password> -h <hostname> --set-gtid-purged=OFF --no-tablespaces --no-create-info <db-name> > data.sql
```

Next, edit `schema.sql` to remove all foreign key constraints ([learn more](https://planetscale.com/docs/learn/operating-without-foreign-key-constraints)), for example:

```diff
CREATE TABLE products (
  id INT NOT NULL,
  category_id INT,
  PRIMARY KEY (id),
+ KEY category_id_idx (category_id)
- KEY category_id_idx (category_id),
- CONSTRAINT `category_fk` FOREIGN KEY (category_id) REFERENCES category(id)
);
```

(watch out for the trailing comma, else you might get errors like "You have an error in your SQL syntax")

Finally, you can import the `schema.sql` and `data.sql` into PlanetScale, **using the PlanetScale settings this time** (user, password, host):

```bash
mysql -u <user> -p<password> -h <hostname> <db-name> < schema.sql
mysql -u <user> -p<password> -h <hostname> <db-name> < data.sql
```

## Schema changes workflow

PlanetScale has a concept of [database branches](https://planetscale.com/docs/concepts/branching):

> - [**Development** branches](https://planetscale.com/docs/concepts/branching#development-and-production-branches) accept schema changes.

Reviewer: ```suggestion
- [**Development** branches](https://planetscale.com/docs/concepts/branching#development-and-production-branches) are isolated copies of your production database and are used to test schema changes in development.
```

> - [**Production** branches](https://planetscale.com/docs/concepts/branching#development-and-production-branches) are built for scale and high availability. Schema changes are forbidden.

Reviewer: ```suggestion
- [**Production** branches](https://planetscale.com/docs/concepts/branching#development-and-production-branches) are high availability branches intended for production traffic. They are protected from direct DDL, so you cannot perform direct schema changes on production branches.
```

Author: Oh I really like that wording better!

- [**Development** branches](https://planetscale.com/docs/concepts/branching#development-and-production-branches) are isolated copies of your production database and are used to test schema changes in development.
- [**Production** branches](https://planetscale.com/docs/concepts/branching#development-and-production-branches) are high availability branches intended for production traffic. They are protected from direct DDL, so you cannot perform direct schema changes on production branches.

You can set up production branches in two ways:

> - either **allow** running DB migrations in production,

Author: 💯 will do!
